### PR TITLE
Add multiple rules, update test structure, and require Node.js 8

### LINKED ISSUE #4: Table of Contents rule
Would be useful to have a rule that ensures a correct `Contents` (TOC):
- Ensure the first section is named `Contents`.
- Ensure all the links point to existing sections with the same name.
- Ensure all sections are included except for the `License` section.
- Ensure the `Contents` list is maximum 2 levels. (People have a tendency of creating too big TOCs)

### LINKED ISSUE #5: List item style rule
The style is:

```
- [Name](Link) - Description.
```

The rule should:
- Ensure the link and description are delimited with `-`.
- Ensure the description starts with an uppercase.
- Ensure the description ends in a dot `.`.

### LINKED ISSUE #6: License rule
Need a rule to ensure a license is included and that it's one of the Creative Commons licenses.

https://github.com/sindresorhus/awesome/blob/master/awesome.md#choose-an-appropriate-license

### LINKED ISSUE #7: Contribution guidelines rule
Should ensure there's a contributing.md file in the repo with some content.

### LINKED ISSUE #23: "Missing Awesome badge": False positive?
awesome-lint reports
```
✖  1:1  Missing Awesome badge after the main heading  awesome/badge
```
for [my list](https://github.com/nschloe/awesome-numerical-software/) which starts off with
```markdown
# Awesome Numerical Software [![Awesome](https://awesome.re/badge-flat.svg)](https://github.com/sindresorhus/awesome)
```
not sure what's wrong here.

### PR DESCRIPTION
**Pull request by transitive-bullshit**
This PR combines #24, #26, #27, and #28 because it was getting too difficult to make progress with every individual change in separate branches.

## awesome/license

Fixes #6.

- [x] Ensure that a section named `License` exists with depth 2.
- [x] Ensure that `License` is the last section in the document.
- [x] Ensure that `License` is not empty.
- [x] Ensure that any `License` image that exists is not png.

This PR doesn't do any additional license validation against CC0 or any other licenses. If desired, we can add that in a future pass.

## awesome/list-item

Fixes #5. Took inspiration from #9.

This PR won't enforce **all** cases that we'll likely want to enforce, but it should enforce most cases with the goal of minimizing false errors.

## awesome/badge

Fixes #23.  I also think this is just generally a good change to make since the templates used to create new awesome lists do not match the current `awesome/badge` URL rules.

## awesome/toc

Fixes #4.

- [x] Ensure the first section is named Contents.
- [x] Ensure all the links point to existing sections with the same name.
- [x] Ensure all sections are included except for the License section.
- [x] Ensure the Contents list is maximum 2 levels.
- [x] Ensure the sections follow the order of the table.
- [x] Ensure sub-sections align up to 2 levels deep (this limit is easily configurable).
  - Note that if a sub-list is present in the ToC, then we make sure its contents match the sub-sections in the body, but if a sub-list is **not** present, then we ignore any possible sub-sections in the body.

This PR won't handle all possible TOC issues, but it should handle the majority of them.

@sindresorhus @wooorm note that instead of breaking this implementation out into a dozen or so individual rules as suggested in #4, I chose to KISS for now. In the short term, it'll be a lot easier for development and maintenance to have them all implemented in one place than spread across multiple files or packages. If we find this functionality generally useful, we can always break it out into more general, encapsulated rules in the future.

## awesome/no-ci-badge

This rule makes sure there are no travis-ci or circleci badges in the readme.

## awesome/contributing

Fixes #7.

This rule makes sure there is a `contributing.md` file (case-insensitive) in the same directory as the readme.

## awesome/git-repo-age

This rule makes sure that the readme resides in a `git` repo whose first commit was at least 30 days ago (this minimum age limit is easily configurable).

cc @matheuss @sindresorhus @wooorm

### CODE REVIEW

**Comment by transitive-bullshit** [comment]
@sindresorhus I've gone through ~15 submission PRs  to awesome, and this version of `awesome-lint` is catching a lot of very valid errors. Out of those 15, only two passed linting.

imho this validates the utility of this type of automated linting. Most people don't follow basic guidelines that this type of linting can easily check 💯

File: .travis.yml [hunk 1]
```diff
@@ -2,4 +2,3 @@ language: node_js
 node_js:
   - '10'
   - '8'
-  - '6'
```
Comment (transitive-bullshit) [inline .travis.yml]: I added some async / await and see no reason to maintain support for non-async-await versions of Node.js. This is, after all, a CLI aimed at open source maintainers.

File: lib/find-readme-file.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+'use strict';
```
Comment (transitive-bullshit) [inline lib/find-readme-file.js]: this logic follows GitHub's precedence rules for readmes. It should probably be broken out into its own npm package eventually.

File: test/rules/contributing.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	const messages = await m({config, filename: 'test/fixtures/contributing/error1/readme.md'});
+	t.deepEqual(messages, [
+		{
+			ruleId: 'awesome/contributing',
+			message: 'contributing.md file must not be empty'
+		}
+	]);
+});
+
+// TODO: this test works locally but fails on CI; @sindresorhus any idea why globby is failing here?
```
Comment (sindresorhus) [inline test/rules/contributing.js]: In what way is it failing?

**Comment by sindresorhus** [comment]
```
❯ awesome-lint https://github.com/sindresorhus/awesome-electron

  ../../../../../private/var/folders/3x/jf5977fn79jbglr7rk0tq4d00000gn/T/8ba94a0960f2e9d5c4b9f40035b33ee2/readme.md:100:65
  ✖  100:65   List item description must start with a non-empty string      awesome/list-item
  ✖  160:38   List item description must start with valid casing            awesome/list-item
  ✖  175:57   List item description must end with proper punctuation        awesome/list-item
  ✖  181:52   List item description must end with proper punctuation        awesome/list-item
  ✖  200:72   List item description must end with proper punctuation        awesome/list-item
  ✖  201:69   List item description must end with proper punctuation        awesome/list-item
  ✖  202:68   List item description must end with proper punctuation        awesome/list-item
  ✖  203:64   List item description must end with proper punctuation        awesome/list-item
  ✖  235:79   List item link and description must be separated with a dash  awesome/list-item
  ✖  237:73   List item link and description must be separated with a dash  awesome/list-item
  ✖  296:74   List item description must start with a non-empty string      awesome/list-item
  ✖  365:231  List item description must end with proper punctuation        awesome/list-item
  ✖  366:230  List item description must end with proper punctuation        awesome/list-item
  ✖  367:107  List item link and description must be separated with a dash  awesome/list-item
  ✖  380:106  List item link and description must be separated with a dash  awesome/list-item
  ✖  381:161  List item link and description must be separated with a dash  awesome/list-item
  ✖  383:126  List item link and description must be separated with a dash  awesome/list-item
  ✖  384:118  List item link and description must be separated with a dash  awesome/list-item
  ✖  403:3    Invalid list item link                                        awesome/list-item
  ✖  404:3    Invalid list item link                                        awesome/list-item
  ✖  405:3    Invalid list item link                                        awesome/list-item
  ✖  406:59   List item link and description must be separated with a dash  awesome/list-item
  ✖  407:47   List item link and description must be separated with a dash  awesome/list-item
  ✖  408:50   List item link and description must be separated with a dash  awesome/list-item
  ✖  409:46   List item link and description must be separated with a dash  awesome/list-item
  ✖  410:3    Invalid list item link                                        awesome/list-item
  ✖  411:3    Invalid list item link                                        awesome/list-item

  27 errors
```

Many of these are false-positives and some should be supported, like having a link in a description.

File: rules/toc.js [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	});
+
+	return links;
+}
+
+function validateListItems({ast, file, list, headingLinks, headings, depth}) {
+	let index = 0;
+
+	if (list) {
+		for (; index < list.children.length; ++index) {
```
[the remark this answers is not in the document]
Comment (transitive-bullshit) [reply rules/toc.js]: I am using the index after the for loop to tell if there are any headings that don't exist in children, so this wouldn't work as-is.

File: rules/toc.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+				}
+			} else {
+				// No need to enforce the existence of a subList, even if there are corresponding
+				// subHeadings.
+			}
+		}
+	}
+
+	if (index < headings.length) {
+		for (; index < headings.length; ++index) {
```
[the remark this answers is not in the document]
Comment (transitive-bullshit) [reply rules/toc.js]: This is not intended to loop over all headings, but rather only the extra headings that are left after we traverse all list children.

**Comment by transitive-bullshit** [comment]
> Many of these are false-positives and some should be supported, like having a link in a description.

There are definitely some false positive cases that we need to weed out and add to the list of unit tests that should pass, especially for `awesome/list-item`. With that being said, things like having a link in the description are already [supported](https://github.com/transitive-bullshit/awesome-lint/blob/feature/up-to-date/test/fixtures/list-item/0.md).

File: test/rules/contributing.js (hunk 3 above)
Comment (transitive-bullshit) [reply test/rules/contributing.js]: It was failing to find the "CONTRIBUTING.md" file on travis-ci which meant that globby wasn't finding the file in a case-insensitive fashion, although it works locally for me on Mac OS.

I'll retry and see if the `case: false` option has solved this issue.

**Comment by transitive-bullshit** [comment]
```
awesome-lint https://github.com/sindresorhus/awesome-electron
```

Working through these false positives now.

**Comment by transitive-bullshit** [comment]
```
> awesome-lint https://github.com/sindresorhus/awesome-electron

  readme.md
  ✖  160:38  List item description must start with valid casing            awesome/list-item
  ✖  235:79  List item link and description must be separated with a dash  awesome/list-item
  ✖  237:73  List item link and description must be separated with a dash  awesome/list-item

  3 errors
```

I think these 3 remaining errors are valid. For L160, `npm => NPM` and L235/L235, move the `(by XX)` to the end of the description and it will pass the linter.

**Comment by transitive-bullshit** [comment]
> The `../../../../../private/var/folders/3x/jf5977fn79jbglr7rk0tq4d00000gn/T/8ba94a0960f2e9d5c4b9f40035b33ee2/readme.md:100:65` part is slightly bothering me. Would be nice if we just hid it, or just show the filename part `readme.md:100:65`.

Done.

**Comment by sindresorhus** [comment]
> For L160, npm => NPM

npm is written in lowercase. There are other brand words that also starts with lowercase, like `iOS`, `macOS`, etc. Maybe we should hard-code a whitelist of some of the most popular ones?

**Comment by sindresorhus** [comment]
> move the (by XX) to the end of the description and it will pass the linter.

Agreed. Done: https://github.com/sindresorhus/awesome-electron

**Comment by sindresorhus** [comment]
Awesome Lint helped me catch a missing TOC entry 👌: https://github.com/sindresorhus/awesome-observables/commit/b2d678757eea7d1249e69260e8ae252a6e018a91

The error could have been clearer though:

```
  ✖  15:1   ToC item "Articles" does not match corresponding heading "Libraries"  awesome/toc
```

It should instead have said: `No matching section for ToC item "Libraries"`.

**Comment by sindresorhus** [comment]
I ran it on Awesome itself:

```
❯ awesome-lint https://github.com/sindresorhus/awesome

  readme.md:88:72
  ✖   50:1    First heading should be name of awesome list                  awesome/toc
  ✖   88:72   List item link and description must be separated with a dash  awesome/list-item
  ✖   90:58   List item link and description must be separated with a dash  awesome/list-item
  ✖  108:57   List item link and description must be separated with a dash  awesome/list-item
  ✖  115:69   List item link and description must be separated with a dash  awesome/list-item
  ✖  130:64   List item link and description must be separated with a dash  awesome/list-item
  ✖  133:59   List item link and description must be separated with a dash  awesome/list-item
  ✖  158:57   List item link and description must be separated with a dash  awesome/list-item
  ✖  164:54   List item link and description must be separated with a dash  awesome/list-item
  ✖  183:65   List item link and description must be separated with a dash  awesome/list-item
  ✖  212:4    List item description must end with proper punctuation        awesome/list-item
  ✖  253:73   List item link and description must be separated with a dash  awesome/list-item
  ✖  256:4    List item description must end with proper punctuation        awesome/list-item
  ✖  275:77   List item link and description must be separated with a dash  awesome/list-item
  ✖  277:86   List item link and description must be separated with a dash  awesome/list-item
  ✖  281:108  List item link and description must be separated with a dash  awesome/list-item
  ✖  290:4    List item description must end with proper punctuation        awesome/list-item
  ✖  367:4    List item description must end with proper punctuation        awesome/list-item
  ✖  449:66   List item link and description must be separated with a dash  awesome/list-item
  ✖  476:4    List item description must end with proper punctuation        awesome/list-item
  ✖  501:4    List item description must end with proper punctuation        awesome/list-item
  ✖  526:4    List item description must end with proper punctuation        awesome/list-item
  ✖  536:4    List item description must end with proper punctuation        awesome/list-item
  ✖  584:93   List item link and description must be separated with a dash  awesome/list-item

  24 errors
```


>   ✖   50:1    First heading should be name of awesome list                  awesome/toc

If the first section is `Contents`, it means the header is HTML or something else and should be ignored.

>   ✖   88:72   List item link and description must be separated with a dash  awesome/list-item

There are lots of these errors when an item is missing a description. I think it would be better to clearly state that it's missing a description and have a separate error if it has a description, but the formatting of it is invalid.

>   ✖  212:4    List item description must end with proper punctuation        awesome/list-item

Not sure why indented items gets a different error when missing a description.

**Comment by transitive-bullshit** [comment]
> There are lots of these errors when an item is missing a description. I think it would be better to clearly state that it's missing a description and have a separate error if it has a description, but the formatting of it is invalid.

Agreed; currently, these are failing because they're list-items that contain sub-lists which I need to add support for.

**Comment by transitive-bullshit** [comment]
Using my forked version of `remark` that supports parsing sub-lists indented with tabs, most of these noise errors go away. There are now 3 remaining errors when linting `awesome`, all of which I believe are valid:

```
❯ awesome-lint https://github.com/sindresorhus/awesome

  readme.md:133:59
  ✖  133:59  List item link and description must be separated with a dash    awesome/list-item
  ✖  449:66  List item link and description separated by invalid whitespace  awesome/list-item
  ✖  584:93  List item link and description separated by invalid en-dash     awesome/list-item

  3 errors
```

`L133` is erroring because the sub-list uses mixed indentation (spaces and tabs) which does not appear to be parsing correctly with any version of remark, though it does render correctly on GitHub. This isn't a great error at the moment, but it is a reasonable thing to warn about.

`L449` is a valid error imho because it's using a unicode `&nbsp;` (non-breaking space) instead of a normal ascii space ` `.

`L584` is a valid error imho because it's using a unicode `—` (en-dash) instead of a normal ascii dash `-`.

**Comment by transitive-bullshit** [comment]
> There are lots of these errors when an item is missing a description. I think it would be better to clearly state that it's missing a description and have a separate error if it has a description, but the formatting of it is invalid.

These errors weren't because of missing descriptions. We do not require any description for list-items. We are requiring, however, that if you include a description it should be formatted properly.

These errors are due to remark not recognizing the sub-lists and instead including them as part of the description text. My [fix for this issue](https://github.com/remarkjs/remark/pull/347) will likely take some time to land.

@sindresorhus I would recommend that we merge this PR once you're satisfied and follow-up with smaller, more focused PRs for these types of issues, but at the end of the day it's up to you 💯 We don't have to start promoting `awesome-lint` usage until we're all fully satisfied that it's up to par.

**Comment by sindresorhus** [comment]
> There are now 3 remaining errors when linting awesome, all of which I believe are valid:

👍 https://github.com/sindresorhus/awesome/commit/1f43f0d6895da786c275c7c18e794c6d17b16c5b

**Comment by sindresorhus** [comment]
I'm very happy with where we're at now. I'm sure there are lots of more things that need to be done, but this is a huge improvement!

**Comment by transitive-bullshit** [comment]
Awesome -- thanks @sindresorhus! 💯

**Pull request merged**